### ISSUE TITLE
Conflicting definitions for 'node' found

### ISSUE THREAD

**Issue by mgroenhoff** [electron/electron#10167]
Currently (1.6.9 - 1.7.5) there is a dependency on @types/node with version ^7.0.18 in package.json. If you have lets say version 8.x.x as one of your own dependencies you get the following compile error:

> .../node_modules/electron/electron.d.ts(6,23): message TS4090: Conflicting definitions for 'node' found at '.../node_modules/electron/node_modules/@types/node/index.d.ts' and '.../node_modules/@types/node/index.d.ts'. Consider installing a specific version of this library to resolve the conflict.

I am now stuck with using v1.6.8 and typings from DefinitelyTyped

**Comment by felixrieseberg** [comment]
Let's make sure I understand you right - you're including your own Node typings? Given that Electron comes with its own Node runtime, doesn't it make sense for Electron to include the correct version?

**Comment by mgroenhoff** [comment]
Yes i currently use part of the same codebase for an application that runs on the server and some that runs in an electron process. I have installed electron as a regular dependency and it also comes with it's own typings hence the conflicting versions in the compile process.

**Comment by zeke** [comment]
What happens if you `rm -rf node_modules/electron/node_modules/@types/node` ?

**Comment by MarshallOfSound** [comment]
@zeke That would fix the problem but it's probably not what we want to do, we still want to ship the "correct" version of the node typings for Electron

@mgroenhoff So you are sharing a single `node_modules` folder for both an Electron application and a server application?  I'd suggest if that project structure is definitely what you want to be doing you should remove your own dependency on `@types/node`.  The whole point of the typings file is to strictly document the methods / return types that are available to you.  If you are using the `v8.x` version of the typings file with Electron it will be incorrect and could lead to valid typescript compilation but then cause runtime errors when you try call methods that don't exist.  You should use the lowest common denominator of methods which would be those documented in `v7.x` of the node typings (the ones we ship).

TLDR: If you want type safety and are using node@8 and electron@1.6/1.7 you should use the typings shipped with electron as those methods will exist in both environments.

**Comment by zeke** [comment]
Not an answer to your question, but worth mentioning: [electron 1.8 will include node 8](https://github.com/electron/electron/pull/9992). We will bump `@types/node` as part of that release.

**Comment by mgroenhoff** [comment]
A lot of projects that bundle their own typings have faced the same issue and they solved that by setting `"@types/node": "*"` in package.json to let the developers choose which typings they use. I understand that in the case of electron this makes no real sense because it is meant to be used inside of an electron process where the node version is fixed.

I have found a workaround:

Copy `<project>/node_modules/electron/electron.d.ts` as is to somewhere outside of your node_modules folder for example `<project>/typings/electron.d.ts`
Add the following to your `tsconfig.json` so Typescript will first search in this directory for typings and then continue with normal resolution. This causes Typescript to resolve the reference to the node typings to my own version.
```
{
	"compilerOptions": {
		"baseUrl": "",
		"paths": {
			"*": [
				"typings/*"
			]
		}
	}
}
```

**Comment by zeke** [comment]
Thanks for following up with your workaround @mgroenhoff 👍

**Comment by ambroiseRabier** [comment]
Have the same error, two definition of @types/node, one from the angular CLI project, that is needed by electron-updater. The other definition of @types/node is from electron itself. The workaround did not work for me, did I miss something? I get this error.

```
node_modules/builder-util-runtime/out/httpExecutor.d.ts(25,39): error TS1039: Initializers are not allowed in ambient contexts.
node_modules/electron-updater/out/AppUpdater.d.ts(12,18): error TS2503: Cannot find namespace 'Electron'.
node_modules/electron-updater/out/electronHttpExecutor.d.ts(4,18): error TS2503: Cannot find namespace 'Electron'.
node_modules/electron-updater/out/electronHttpExecutor.d.ts(5,24): error TS2503: Cannot find namespace 'Electron'.
node_modules/electron-updater/out/electronHttpExecutor.d.ts(9,64): error TS2503: Cannot find namespace 'Electron'.
```

**Comment by dkapp834** [comment]
For me, I had an outdated npm shrinkwrap file.

**Comment by snicol21** [comment]
For what it's worth I fixed very similar errors by changing the way I imported electron-updater
From: `import { autoUpdater } from 'electron-updater';`
To: `const { autoUpdater } = require('electron-updater');`

**Comment by incomingstick** [comment]
@snicol21 your change fixed this error for me. How did you narrow it down to the updater?

**Comment by snicol21** [comment]
@incomingstick lots of trial and error. Wish I had more of a logical explanation, haha. Glad it helped.

**Comment by kunokdev** [comment]
Very sad that electron developers still let this issue persist instead of following convenience like every other lib. 

@mgroenhoff  's workaround works, which is terrible approach btw, but thanks, at least it works after hours of trying everything else.

**Comment by fmacedoo** [comment]
2022 and it still happens!

**Comment by filipenatanael** [comment]
Hi guys! @snicol21 @fmacedoo @incomingstick @kunokdev

Here is an explanation of @snicol21 workaround.

"When exporting a module using _**export =**_, TypeScript-specific _**import module = require("module")**_ must be used to import the module." _Available on typescriptlang.org_

If you have a look inside the **electron-updater**, you are going to find something similar to the pseudocode bellow.

```javascript
// Inside dependency folder...
export = autoUpdater; 

// Your implementation...
import { autoUpdater } from 'electron-updater'; 
```

**_'autoUpdater'_** can only be imported by using '**_const Hello = require("electron-updater")_**' or by turning on the '**esModuleInterop**' flag and using a default import.
